lib/nrf_modem_lib/nrf_modem_lib_trace.c

```diff
@@ -27,9 +39,94 @@ static char rtt_buffer[CONFIG_NRF_MODEM_LIB_TRACE_MEDIUM_RTT_BUF_SIZE];
 
 static bool is_transport_initialized;
 
+struct trace_data_t {
+	void *fifo_reserved; /* 1st word reserved for use by fifo */
+	const uint8_t * const data;
+	uint32_t len;
+};
+
+K_FIFO_DEFINE(trace_fifo);
+
+#define TRACE_THREAD_STACK_SIZE 256
+#define TRACE_THREAD_PRIORITY CONFIG_NRF_MODEM_LIB_TRACE_THREAD_PRIO
+
+void trace_handler_thread(void)
+{
+	while (1) {
+		struct trace_data_t *trace_data = k_fifo_get(&trace_fifo, K_FOREVER);
+		const uint8_t * const data = trace_data->data;
+		const uint32_t len = trace_data->len;
+
 #ifdef CONFIG_NRF_MODEM_LIB_TRACE_MEDIUM_UART
+		/* Split RAM buffer into smaller chunks to be transferred using DMA. */
+		const uint32_t MAX_BUF_LEN = (1 << UARTE1_EASYDMA_MAXCNT_SIZE) - 1;
+		uint32_t remaining_bytes = len;
+		nrfx_err_t err;
+
+		while (remaining_bytes) {
+			size_t transfer_len = MIN(remaining_bytes, MAX_BUF_LEN);
+			uint32_t idx = len - remaining_bytes;
+
+			if (k_sem_take(&tx_sem, K_MSEC(UART_TX_WAIT_TIME_MS)) != 0) {
+				LOG_WRN("UARTE TX not available!");
+				break;
+			}
+			err = nrfx_uarte_tx(&uarte_inst, &data[idx], transfer_len);
```

Comment:
Was wondering is there some reason why using nrfx uart directly and Zephyr async API is not used (uart_tx())?
AFAIK, uart_tx() translates to nrfx_uarte_tx().

Reply:
This is legacy code. But I believe it is because zephyr driver cant be used as the UART1 is disabled in the device overlay here -> https://github.com/nrfconnect/sdk-zephyr/blob/c3208e7ff49d22d8271f305344382e9306fdde99/boards/arm/nrf9160dk_nrf9160/nrf9160dk_nrf9160_ns.dts#L21

Reply:
@jhirsi I wanted to use the Zephyr async API, but it didn't look mature enough when I looked into using it in October last year. Back then it was a completely separate UARTE-implementation and it looks like that is still the case, but I haven't tested it lately. What would be the benefit to use the Zephyr async API?

Reply:
@gregersrygg, Zephyr async uart api seems to be working good nowadays, and actually, just quite recently, implemented support by using it to Zephyr ppp driver for mosh ppp dialup usage. Performance impact was huge for dial up TCP/IP performance, although I compared to existing polling uart in ppp. But anyways, I also started with nrfx uart but uart folks convinced me to use zephyr api. Eventually easier to maintain and I think that also to get rid of the need for that kind of workarounds that are needed due to this in mosh: https://github.com/nrfconnect/sdk-nrf/pull/7080